hub/core/chunk.py

```diff
@@ -0,0 +1,141 @@
+        return len(self._data)
+
+    def is_under_min_space(self, min_data_bytes_target: int):
+        return self.num_data_bytes < min_data_bytes_target
+
+    def has_space_for(self, num_bytes: int, max_data_bytes: int):
+        return self.num_data_bytes + num_bytes <= max_data_bytes
+
+    def append_sample(
+        self, incoming_buffer: memoryview, max_data_bytes: int
+    ) -> Tuple["Chunk"]:
+        """Store `incoming_buffer` in this chunk.
+
+        Args:
+            incoming_buffer (memoryview): Buffer that represents a single sample.
+            max_data_bytes (int): Used to determine if this chunk has space for `incoming_buffer`.
+
+        Raises:
+            FullChunkError: If `incoming_buffer` is too large.
+        """
+
+        incoming_num_bytes = len(incoming_buffer)
+
+        if not self.has_space_for(incoming_num_bytes, max_data_bytes):
+            raise FullChunkError(
+                f"Chunk does not have space for the incoming bytes (incoming={incoming_num_bytes}, max={max_data_bytes})."
+            )
+
+        # note: incoming_num_bytes can be 0 (empty sample)
+        self._data += incoming_buffer
+
+    def update_headers(
+        self, incoming_num_bytes: int, num_samples: int, sample_shape: Sequence[int]
+    ):
+        """Updates this chunk's header. A chunk should NOT exist without headers.
+
+        Args:
+            incoming_num_bytes (int): Number of bytes this header should account for. Should be divisble by `num_samples`.
+            num_samples (int): Number of samples this header should account for.
+            sample_shape (Sequence[int]): Every sample that `num_samples` symbolizes is considered to have `sample_shape`.
+
+        Raises:
+            Exception: If trying to update headers when no data was actually added.
+            ValueError: If `incoming_num_bytes` is not divisible by `num_samples`.
+        """
+
+        if incoming_num_bytes % num_samples != 0:
+            raise ValueError(
+                "Incoming bytes should be divisible by the number of samples to properly update headers."
+            )
+
+        num_bytes_per_sample = incoming_num_bytes // num_samples
+        self.shapes_encoder.add_shape(sample_shape, num_samples)
+        self.byte_positions_encoder.add_byte_position(num_bytes_per_sample, num_samples)
+
+    def __len__(self):
+        # this should not call `tobytes` because it will be slow. should calculate the amount of bytes this chunk takes up in total. (including headers)
+
+        shape_nbytes = self.shapes_encoder.nbytes
+        range_nbytes = self.byte_positions_encoder.nbytes
+        error_bytes = 32  # to account for any extra delimeters/stuff that `np.savez` may create in excess
+
+        return shape_nbytes + range_nbytes + self.num_data_bytes + error_bytes
+
+    def tobytes(self) -> memoryview:
+        out = BytesIO()
+
+        # TODO: for fault tolerance, we should have a chunk store the ID for the next chunk
```

Comment:
still not sold on this

Reply:
didnt ask

Reply:
😝

Reply:
@mccrearyd it's called a pull _request_ for a reason.